### ISSUE TITLE
When the initial element of the tuple is a rest element, it is not possible to correctly infer the type of the initial element.

### ISSUE THREAD

**Issue by 13OnTheCode** [microsoft/TypeScript#56885]
### 🔍 Search Terms

initial element of the tuple is a rest element

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

```ts
type T1 = [...string[], number, boolean]

declare const head: T1[0]
         // Actual: string | number | boolean
         // Expected: string | number
```

### 📃 Motivating Example

-

### 💻 Use Cases

[https://github.com/microsoft/TypeScript/issues/56883](https://github.com/microsoft/TypeScript/issues/56883)

**Comment by craigphicks** [comment]
I actually posted this same issue/proposal [long form](#49105), [short form](#49138).  The latter is labeled as a bug, and is open.
I also have a [pull for it](#49106), which needs to be updated to the latest version and resubmitted.

**Comment by craigphicks** [comment]
@fatcerberus
Let me clarify my statement.  

I am making two assumptions:
1. that the current assignment behavior is correct
2. that correct indexing behavior should follow exactly the results of assignment behavior

@jcalz asserts [here]() that *"Yes, that type (`arr[0]`) is wider than it should be, but that's not a bug, it's a design decision."* 

In other words, that my assumption 2 is incorrect.  

Certainly there are times when economy of calculation dictates using wider types.  Other times it may just be expedient coding.
Either of those would enough to invalidate my assumption 2.   

**Therefore I will withdraw my statement that this must be a bug, and instead assume it is a feature request.**

**Comment by craigphicks** [comment]
@13OnTheCode - As a feature request, I believe more explanation is required.

Currently, you have your *motivating example* code listed in the *suggestion* section, and the *motivating example* and *use cases* sections are empty.

Here is an example:

> ### Suggestion
> - Indexed types should correspond exactly to results of assignment.

You could also give an algorithm that computes that.  

One consideration - if it takes O(N) to compute each indices type individually, where N is the length of the tuple,
then to compute all N types individually would take O(N^2).

However, it should be possible to compute all the indices types in sequence together in O(N) time, and cache the result.

To do that correctly, it is necessary to understand in detail the corner cases of the current assignment algorithm.

**Comment by fatcerberus** [comment]
> ...that correct indexing behavior should follow exactly the results of assignment behavior

Yes, this assumption is incorrect.  The PR that implemented mid/lead rest elements (https://github.com/microsoft/TypeScript/pull/41544) specifically calls out the current indexing behavior, suggesting it's intentional (or at least a known-upfront design limitation).

From #41544:
> Also note that **indexing a tuple type beyond its starting fixed elements** (if any) **yields a union type of all possible element types**
